azimuth/types/model_performance.py

```
@@ -96,9 +103,7 @@ class MetricsPerFilterValue(MetricsResponseCommonFields, UtteranceCountPerFilter
     pass
 
 
-class MetricsPerFilter(
-    ValuePerDatasetFilter[MetricsPerFilterValue], ValuePerPipelineFilter[MetricsPerFilterValue]
-):
+class MetricsPerFilter(ValuePerClassFilter[MetricsPerFilterValue]):
     pass
```

The new class `ValuePerClassFilter` is only used by `MetricsPerFilter`, right?
I find the name `ClassFilter` confusing since it includes (inherits from) the smart tags, which are not classes.
### Solution 1
Remove the smart tags from `ValuePerClassFilter` and add them back to `MetricsPerFilter`. I wonder if that's what you meant to do allalong.
```diff
-class ValuePerClassFilter(
-    ValuePerPipelineSmartTag[T], ValuePerDatasetSmartTag[T], GenericModel, Generic[T]
-):
+class ValuePerClassFilter(GenericModel, Generic[T]):
```
```suggestion
class MetricsPerFilter(
    ValuePerClassFilter[MetricsPerFilterValue],
    ValuePerDatasetSmartTag[MetricsPerFilterValue],
    ValuePerPipelineSmartTag[MetricsPerFilterValue],
):
    pass
```
### Solution 2
Avoid having to name a new class altogether by defining `MetricsPerFilter` directly:
```diff
-class ValuePerClassFilter(
-    ValuePerPipelineSmartTag[T], ValuePerDatasetSmartTag[T], GenericModel, Generic[T]
-):
-    label: List[T] = Field(..., title="Label")
-    prediction: List[T] = Field(..., title="Prediction")
-
-
```
```suggestion
class MetricsPerFilter(
    ValuePerDatasetSmartTag[MetricsPerFilterValue], ValuePerPipelineSmartTag[MetricsPerFilterValue]
):
    label: List[MetricsPerFilterValue] = Field(..., title="Label")
    prediction: List[MetricsPerFilterValue] = Field(..., title="Prediction")
```
### Solution 3
A variation of solution 1 that avoids all duplication. Instead of `ValuePerClassFilter`, create two `ValuePerLabel` and `ValuePerClassPrediction` that can be reused in `ValuePerDatasetFilter` and `ValuePerPrediction`. However, mypy complains about this one and I don't understand why. 😢 
```diff
-class ValuePerDatasetFilter(ValuePerDatasetSmartTag[T], GenericModel, Generic[T]):
+class ValuePerLabel(GenericModel, Generic[T]):
     label: List[T] = Field(..., title="Label")
+
+
+class ValuePerDatasetFilter(ValuePerLabel[T], ValuePerDatasetSmartTag[T], GenericModel, Generic[T]):
     data_action: List[T] = Field(..., title="Data action tag")
 
 
-class ValuePerPipelineFilter(ValuePerPipelineSmartTag[T], GenericModel, Generic[T]):
+class ValuePerPrediction(GenericModel, Generic[T]):
     prediction: List[T] = Field(..., title="Prediction")
+
+
+class ValuePerPipelineFilter(
+    ValuePerPrediction[T], ValuePerPipelineSmartTag[T], GenericModel, Generic[T]
+):
     outcome: List[T] = Field(..., title="Outcome")
-
-
-class ValuePerClassFilter(
-    ValuePerPipelineSmartTag[T], ValuePerDatasetSmartTag[T], GenericModel, Generic[T]
-):
-    label: List[T] = Field(..., title="Label")
-    prediction: List[T] = Field(..., title="Prediction")
```
```suggestion
class MetricsPerFilter(
    ValuePerLabel[MetricsPerFilterValue],
    ValuePerDatasetSmartTag[MetricsPerFilterValue],
    ValuePerPrediction[MetricsPerFilterValue],
    ValuePerPipelineSmartTag[MetricsPerFilterValue],
):
    pass
```

---

Thank you! I was trying to do Solution 2, but I didn't understand well the dynamic with the `[T]` so I couldn't make it work. That's the moment where I asked for your help but then said "well I found something that makes it not crash, so you'll see in the PR if that makes sense". Thanks for your help. :)

---

I have tried Solution 2, but I get a `mypy` error : `azimuth/modules/model_performance/metrics.py:268: error: Too many arguments for "MetricsPerFilter"  [call-arg]`. Do you have any idea?

---

Oups, ok. If you pull your branch, you'll get a commit from me that fixes that. It's not really related, but the current code was hiding the error.

---

Oh awesome, thanks! I didn't see it.

---

Of course! I should have warned you, sorry.